### PR TITLE
Fix DeepSpeed MoE

### PR DESCRIPTION
**Pull request by conglongli**
Fix several issues about DeepSpeed MoE training after recent rebase:

1. Fix bug that num_experts not properly passed when initializing model (bug reported by https://github.com/microsoft/DeepSpeed/issues/4039)
2. Add back args.local_rank which is needed by DeepSpeed
3. Fix an issue in `megatron/core/tensor_parallel/layers.py` where when DeepSpeed MoE is used, the number of dimensions is different from expected (may need further investigation).

After the above fixes, confirmed that DeepSpeed MoE training work fine and the loss curve match the training with older Megatron-DeepSpeed.

### CODE REVIEW

File: megatron/core/tensor_parallel/layers.py [hunk 1]
```diff
@@ -279,10 +279,17 @@ def backward(ctx, grad_output):
...
-                                       grad_output.shape[2])
-        total_input = total_input.view(total_input.shape[0] * total_input.shape[1],
-				       total_input.shape[2])
+        if len(grad_output.shape) == 3:
+            grad_output = grad_output.view(grad_output.shape[0] * grad_output.shape[1],
+                                        grad_output.shape[2])
+            total_input = total_input.view(total_input.shape[0] * total_input.shape[1],
+                        total_input.shape[2])
+        else:
+            # Somehow when DeepSpeed MoE is used, grad_output could have 4 dimensions.
```
Comment (clumsy) [inline megatron/core/tensor_parallel/layers.py:288]: Do we know what these dimensions are? MoE should not add new ones.

File: megatron/arguments.py [hunk 2]
```diff
@@ -1114,6 +1114,8 @@ def _add_distributed_args(parser):
                        'a custom built image that support ring-exchange p2p.')
     group.add_argument('--local-rank', type=int, default=None,
                        help='local rank passed from distributed launcher.')
+    group.add_argument('--local_rank', type=int, default=None,
```
Comment (clumsy) [inline megatron/arguments.py:1117]: Can't we simply add `'--local_rank'` alias on Line#1115?

Comment (conglongli) [reply megatron/arguments.py:1117]: Yeah I just commit a PR to merge them.

File: megatron/core/tensor_parallel/layers.py (hunk 1 above)
Comment (conglongli) [reply megatron/core/tensor_parallel/layers.py:288]: Right now I don't have enough bandwidth for further investigation. But I at least verified that with the above fix, GPT MoE training loss is identical to the case before the whole recent rebase.
